text/0000-const-ub.md

```diff
@@ -0,0 +1,111 @@
+for example, if there is UB while computing the initial value of a `Vec<i32>`, the result might be a completely invalid vector that causes UB at runtime when used in the program.
+
+Sometimes, the compiler might be able to detect such problems and show an error or warning about CTFE computation having gone wrong (for example, the compiler might detect when the array length ends up being uninitialized).
+But other times, this might not be the case -- there is no guarantee that UB is reliably detected during CTFE.
+This can change from compiler version to compiler version: CTFE code that causes UB could build fine with one compiler and fail to build with another.
+(This is in accordance with the general policy that unsound code is not subject to stability guarantees.)
+Implementations are encouraged to perform as many UB checks as they feasibly can, and they are encouraged to document which UB is and is not detected during CTFE and what the consequences of undetected UB can be, but none of this is required.
+
+## CTFE UB-checking in `rustc`
+
+For `rustc` specifically at the time the RFC is written, a lot of UB will actually be detected reliably:
+* Dereferencing dangling pointers.
+* Using an invalid value in an arithmetic, logical or control-flow operation (e.g. using `3` transmuted to a `bool` value in an `if`, or using an uninitialized integer in `+` or `|`).
+* Violating the precondition of an intrinsic (e.g., reaching an `unreachable` or violating the assumptions of `exact_div`).
+
+If any of these errors arise during CTFE, they will currently be reliably detected and a CTFE error will be raised.
+
+Other kinds of UB are ignored, and evaluation continues as if there was no error.
+* Dereferencing unaligned pointers: memory is accessed at the given address even if it is insufficiently aligned.
+* Violating Rust's aliasing rules: memory is read/written even if that violates aliasing guarantees.
+* Producing an invalid value (but not using it in one of the ways defined above): evaluation continues despite the fact that an invalid value was produced.
+
+`rustc` also currently makes no attempt at detecting library UB.
+
+No UB-exploiting MIR optimizations are currently being performed for CTFE, so a CTFE execution currently will never go wrong in arbitrary ways: UB is either detected, or evaluation continues in a well-defined manner as described above.
+
+However, this is just a snapshot of what `rustc` currently does.
+None of this is *guaranteed*, and `rustc` may relax or otherwise change its UB checking any time.
+
+[UB]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When UB arises as part of CTFE, the result of this evaluation is an unspecified constant, i.e., it is arbitrary, and might not even be of the right type.
```

Comment:
I don't understand what "might not even be of the right type" means.

Reply:
"might not be valid for the result type" is a better way to put it

Reply:
Yes, that is much better. The constant's type field will never be affected by CTFE (at least not by the execution that evaluates the constant's value).